### PR TITLE
WL 21341 DodgeBall

### PR DESCRIPTION
**Pull request by MidnightRift**
Info
- HMD Required.
- Two accounts or people required

# QA

## Setup
 - Create a few spheres that are dynamic and have gravity (-9.8).
- add this script to the spheres: https://raw.githubusercontent.com/MidnightRift/hifi/21341/unpublishedScripts/marketplace/dodgeBall/dodgeBall.js

## Steps - Two Person

| | Steps | Expected results |
| --- | --- | --- |
| 1. | Check that the ball can only be far grabbed from less than 2 meters away  | You should only be able to grab the sphere when you are less then 2 meters away (the lasers not going away is a known bug) |
| 2. | Grab and throw one of the spheres you just created at another avatar | The Sphere should leave your hand as thrown. |
| 3. | Observe while in flight the ball is emitting a trail of particles | You should see a trail of particles emitted behind the ball |
| 4. | When the ball is within 1 meter of your target it should stop (this is a hit) | ball is stopped and stationary in air for a moment. |
| 5. | Observe the trail particles stopped. | The sphere is no longer emitting the trailing particles. |
| 6. | Observe a new particle effect. an explosion of sorts (should last no more then 2 seconds.) | You should see particles emitting from the center outwards to signify an explosion effect, lasting less then 2 seconds. |
| 7. | Observe the ball falling to the ground (its only meant to be stationary for about half a second) |  The ball falls to the ground |

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Can one of the admins verify this patch?

**Comment by themelissabrown** [comment]
add to whitelist

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2609/HighFidelity-Beta-PR11213-7da4bc04d3a9eb4db5ffbd9a30a7e20f97dce57a.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2609/HighFidelity-Beta-PR11213-7da4bc04d3a9eb4db5ffbd9a30a7e20f97dce57a.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2614/HighFidelity-Beta-PR11213-397d2a68534cea8d5c5e832571b676148d223fc9.dmg

**Comment by marko8904** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2618/HighFidelity-Beta-PR11213-397d2a68534cea8d5c5e832571b676148d223fc9.dmg

**Comment by cain-kilgore** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2622/HighFidelity-Beta-PR11213-9ff68ce10c47c2831d8207b58ec1dc4d4cedf7d1.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2622/HighFidelity-Beta-PR11213-9ff68ce10c47c2831d8207b58ec1dc4d4cedf7d1.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2623/HighFidelity-Beta-PR11213-6892cdb57de04c936a73f890119951a5e258c781.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2623/HighFidelity-Beta-PR11213-6892cdb57de04c936a73f890119951a5e258c781.exe

**Comment by cain-kilgore** [comment]
Testing w/Aaron.

**Comment by cain-kilgore** [comment]
Steps | Expected results | Result
-- | -- | ---
Check that the ball can only be far grabbed from less than 2 meters away | You should only be able to grab the sphere when you are less then 2 meters away (the lasers not going away is a known bug) | See notes
Grab and throw one of the spheres you just created at another avatar | The Sphere should leave your hand as thrown. | Pass
Observe while in flight the ball is emitting a trail of particles | You should see a trail of particles emitted behind the ball | Intermittent (See details)
When the ball is within 1 meter of your target it should stop (this is a hit) | ball is stopped and stationary in air for a moment. | Pass
Observe the trail particles stopped. | The sphere is no longer emitting the trailing particles. | Pass
Observe a new particle effect. an explosion of sorts (should last no more then 2 seconds.) | You should see particles emitting from the center outwards to signify an explosion effect, lasting less then 2 seconds. | Pass
Observe the ball falling to the ground (its only meant to be stationary for about half a second) | The ball falls to the ground | Pass

## Test Notes

We were having issues with the Particle Effects that the ball was putting off - it seemed to be very inconsistent and didn't always show up for either Avatar. 

The Particle Effects didn't appear to scale according to the size of the ball, and the 2m requirement seems to be from the center of the ball and not necessarily the edge of the ball.

Additionally, Aaron found a bug which he was able to reproduce, see steps below - he was able to Grip his Controllers and control the ball from anywhere. 
1. Throw the Ball at an Avatar
2. Observe the Ball Stopping in Mid Air
3. Grab the Ball while it's stopped, and you can keep letting go and grabbing it from different places.

###### Intel i5 7600K @ 4Ghz, 16GB DDR4 2133Mhz Corsair Vengeance, Asus Strix GTX 1050Ti 3GB, HTC Vive

**Comment by MidnightRift** [comment]
We were having issues with the Particle Effects that the ball was putting off - it seemed to be very inconsistent and didn't always show up for either Avatar. - **This sounds like LOD issues**

The Particle Effects didn't appear to scale according to the size of the ball, and the 2m requirement seems to be from the center of the ball and not necessarily the edge of the ball. - **There was no test for resizing the ball don't know why you tested that. The requirement for the 2 meters is arbitrary, its meant to stop longer distance grabs 2 M to the core is fine**

Additionally, Aaron found a bug which he was able to reproduce, see steps below - he was able to Grip his Controllers and control the ball from anywhere. - **Can you be more specific. Are you being Forced to release the ball or can you freely move it? are you near grabbing or far grabbing ?**

**Comment by YuppyPlays** [comment]
@MidnightRift 
Regarding the bug, repro:
(This appears to happen at random but here are the steps that seemed to sometimes trigger it)
 - Far-grab the sphere as it's in the "Stop" phase when hitting a player (Within ~ 2 meters of course)
 - Whilst held in far-grab, bring closer to your own avatar until it clips to your hand when brought close enough
 - Throw sphere/ball
 - Press grips (Vive), it will automatically control the object no matter where it is relative to player 
 - whilst using grips controlling the sphere, the player doing this is able to manipulate its movements drastically by rotating hands, moving hands away or towards the body has no effect other than 1:1 movement, that is to say, it is unlike far-grabbing mechanics where it can be brought closer or further away via hand position.

Hope this helps :)

File: unpublishedScripts/marketplace/dodgeBall/dodgeBall.js [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            alphaSpread: 0,
+            alphaStart: 1,
+            alphaFinish: 0,
+            polarStart: 0,
+            polarFinish: 0,
+            azimuthStart: -180,
+            azimuthFinish: 180
+        };
+
+            particleTrailEntity = Entities.addEntity(props);
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/dodgeBall/dodgeBall.js]: style: indentation

File: unpublishedScripts/marketplace/dodgeBall/dodgeBall.js [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+    };
+    this.startNearGrab = function (thisEntityID, triggerHandAndAvatarUUIDArray) {
+        clearProxCheck();
+    };
+
+    this.releaseGrab = function (thisEntityID) {
+
+        if (particleTrailEntity === null) particleTrail();
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/dodgeBall/dodgeBall.js]: 3.4.2 style - conditionals should be on separate line in brackets

File: unpublishedScripts/marketplace/dodgeBall/dodgeBall.js [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        }
+    }
+
+    this.startDistanceGrab = function (thisEntityID, triggerHandAndAvatarUUIDArray) {
+        clearProxCheck();
+        var triggerHand = triggerHandAndAvatarUUIDArray[0];
+        var avatarUUID = triggerHandAndAvatarUUIDArray[1];
+
+        var ballPos = Entities.getEntityProperties(_entityID, ['position']).position;
+        if (Vec3.distance(ballPos, AvatarList.getAvatar(avatarUUID).position) > 2) {
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/dodgeBall/dodgeBall.js]: Magic number 2 - should be named constant

File: unpublishedScripts/marketplace/dodgeBall/dodgeBall.js [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+
+    function clearProxCheck() {
+        if (proxInterval) {
+            Script.clearInterval(proxInterval);
+            Entities.deleteEntity(particleTrailEntity);
+            particleTrailEntity = null;
+        }
+
+        if (proxTimeout) Script.clearTimeout(proxTimeout);
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/dodgeBall/dodgeBall.js]: 3.4.2 style - conditionals should be on separate line in brackets

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2641/HighFidelity-Beta-PR11213-9185e1d80d23ea1fe4f0b70ca02df80f837c98b1.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2641/HighFidelity-Beta-PR11213-9185e1d80d23ea1fe4f0b70ca02df80f837c98b1.exe

**Comment by themelissabrown** [comment]
@MidnightRift @YuppyPlays that bug of being able to release the object and still control it sounds like the same thing we used to see in QA testing a lot. Is it different from when you would take off the glove and still be able to control it?

**Comment by themelissabrown** [comment]
@misslivirose this needs switched to the content repro. Do you have time to walk me through that process and I can do it?

**Comment by misslivirose** [comment]
@themelissabrown Since it's already approved here, feel free to merge. I'll be migrating the unpublished scripts content over sometime in the next week or two.

**Pull request merged**